packages/apollo-gateway/src/index.ts

```diff
@@ -300,12 +300,30 @@ export class ApolloGateway implements GraphQLService {
   }
 
   public async load(options?: { engine?: GraphQLServiceEngineConfig }) {
-    await this.updateComposition(options);
+    if (options && options.engine) {
+      if (!options.engine.graphVariant)
+        this.logger.warn('No graph variant provided. Defaulting to `current`.');
+      this.engineConfig = options.engine;
+    }
 
     if (isManagedConfig(this.config) || this.experimental_pollInterval) {
+      await this.updateComposition();
       if (!this.pollingTimer) this.pollServices();
+    } else {
+      try {
+        await this.updateComposition();
+      } catch {
+        // In the case that we're neither managed nor polling, the gateway should
+        // crash in the event that it can't load service definitions. Leaving the
+        // gateway running in a non-operable state doesn't make sense and this
+        // allows container managers like K8s to restart the broken container.
+        // Note: the error caught within `this.updateComposition` handles logging
+        // for the relevant error.
+        process.exit(1);
```

Comment:
At the very least, I think we need to guard if `process` exists before calling `process.exit`, but as a larger hesitation, I'm worried about to shipping this `process.exit` invocation either way.

Concretely: We have no idea what else is running successfully in this process that we're `exit`ing.   
 It may be doing other important things!  I think providing the primitives to users to make these decisions on their own is fine.  In that regard, a failing health check seems like sufficient criteria for the container system to make a determination as to the process' fate.  If the health check fails, K8s can terminate the process and restart it, no?

Reply:
Agreed on all accounts! Resolved by afa6746.